File: Bulk/Logic/DailyStats.swift
import Foundation

/// One day's nutrition summary used by Progress charts, streaks, and insights.
struct DaySummary: Equatable {
    var dayKey: Date
    var calories: Decimal
    var protein: Decimal

    func calorieGoalReached(minimum: Decimal) -> Bool { calories >= minimum }
    func proteinGoalReached(minimum: Decimal) -> Bool { protein >= minimum }
    func bothGoalsReached(calorieMin: Decimal, proteinMin: Decimal) -> Bool {
        calorieGoalReached(minimum: calorieMin) && proteinGoalReached(minimum: proteinMin)
    }
}

enum DailyStats {
    /// Groups log entries into per-day summaries, sorted ascending by day.
    /// Days without entries are simply absent.
    static func summaries(entries: [LogEntry]) -> [DaySummary] {
        Dictionary(grouping: entries, by: \.dayKey)
            .map { day, dayEntries in
                let totals = NutritionCalculator.dayTotals(entries: dayEntries)
                return DaySummary(dayKey: day, calories: totals.calories, protein: totals.protein)
            }
            .sorted { $0.dayKey < $1.dayKey }
    }

    /// Percentage (0...100) of the given days on which `predicate` holds.
    static func percentage(of days: [DaySummary], where predicate: (DaySummary) -> Bool) -> Double {
        guard !days.isEmpty else { return 0 }
        let hit = days.filter(predicate).count
        return Double(hit) / Double(days.count) * 100
    }

    static func averageCalories(_ days: [DaySummary]) -> Decimal {
        guard !days.isEmpty else { return 0 }
        return days.reduce(Decimal(0)) { $0 + $1.calories } / Decimal(days.count)
    }

    static func averageProtein(_ days: [DaySummary]) -> Decimal {
        guard !days.isEmpty else { return 0 }
        return days.reduce(Decimal(0)) { $0 + $1.protein } / Decimal(days.count)
    }
}

/// Streaks of consecutive calendar days on which both minimums were reached.
/// A missing day (no entries at all) breaks a streak.
enum StreakCalculator {
    static func streaks(
        days: [DaySummary],
        calorieMin: Decimal,
        proteinMin: Decimal,
        today: Date = DayKey.today(),
        calendar: Calendar = .current
    ) -> (current: Int, longest: Int) {
        let hitDays = Set(
            days
                .filter { $0.bothGoalsReached(calorieMin: calorieMin, proteinMin: proteinMin) }
                .map { calendar.startOfDay(for: $0.dayKey) }
        )
        guard !hitDays.isEmpty else { return (0, 0) }

        // Longest run of consecutive hit days anywhere in history.
        var longest = 0
        for day in hitDays {
            let previous = calendar.date(byAdding: .day, value: -1, to: day)!
            guard !hitDays.contains(previous) else { continue } // only start counting at run starts
            var length = 1
            var cursor = day
            while true {
                let next = calendar.date(byAdding: .day, value: 1, to: cursor)!
                guard hitDays.contains(next) else { break }
                length += 1
                cursor = next
            }
            longest = max(longest, length)
        }

        // Current streak counts back from today; an unfinished today doesn't
        // break it — the streak then counts back from yesterday.
        var current = 0
        var cursor = calendar.startOfDay(for: today)
        if !hitDays.contains(cursor) {
            cursor = calendar.date(byAdding: .day, value: -1, to: cursor)!
        }
        while hitDays.contains(cursor) {
            current += 1
            cursor = calendar.date(byAdding: .day, value: -1, to: cursor)!
        }

        return (current, longest)
    }
}

File: web/src/logic/dailyStats.ts
/* Ported from Bulk/Logic/DailyStats.swift: per-day nutrition summaries used
   by Progress charts, streaks, and insights. */

import { shiftDayKey, todayKey } from "../models/dayKey";
import type { LogEntry } from "../models/types";
import { dayTotals } from "./nutrition";

export interface DaySummary {
  dayKey: string;
  calories: number;
  protein: number;
}

export function bothGoalsReached(day: DaySummary, calorieMin: number, proteinMin: number): boolean {
  return day.calories >= calorieMin && day.protein >= proteinMin;
}

/** Groups log entries into per-day summaries, sorted ascending by day.
    Days without entries are simply absent. */
export function daySummaries(entries: LogEntry[]): DaySummary[] {
  const byDay = new Map<string, LogEntry[]>();
  for (const e of entries) {
    const list = byDay.get(e.dayKey) ?? [];
    list.push(e);
    byDay.set(e.dayKey, list);
  }
  return [...byDay.entries()]
    .map(([dayKey, dayEntries]) => {
      const totals = dayTotals(dayEntries);
      return { dayKey, calories: totals.calories, protein: totals.protein };
    })
    .sort((a, b) => (a.dayKey < b.dayKey ? -1 : 1));
}

/** Percentage (0...100) of the given days on which `predicate` holds. */
export function percentageOfDays(
  days: DaySummary[],
  predicate: (d: DaySummary) => boolean,
): number {
  if (days.length === 0) return 0;
  return (days.filter(predicate).length / days.length) * 100;
}

export function averageCalories(days: DaySummary[]): number {
  if (days.length === 0) return 0;
  return days.reduce((sum, d) => sum + d.calories, 0) / days.length;
}

export function averageProtein(days: DaySummary[]): number {
  if (days.length === 0) return 0;
  return days.reduce((sum, d) => sum + d.protein, 0) / days.length;
}

/** Current and longest streaks of days hitting both goals. An unfinished
    today doesn't break the current streak — it then counts from yesterday. */
export function streaks(
  days: DaySummary[],
  calorieMin: number,
  proteinMin: number,
  today: string = todayKey(),
): { current: number; longest: number } {
  const hitDays = new Set(
    days.filter((d) => bothGoalsReached(d, calorieMin, proteinMin)).map((d) => d.dayKey),
  );
  if (hitDays.size === 0) return { current: 0, longest: 0 };

  let longest = 0;
  for (const day of hitDays) {
    if (hitDays.has(shiftDayKey(day, -1))) continue; // only start counting at run starts
    let length = 1;
    let cursor = day;
    while (hitDays.has(shiftDayKey(cursor, 1))) {
      length += 1;
      cursor = shiftDayKey(cursor, 1);
    }
    longest = Math.max(longest, length);
  }

  let current = 0;
  let cursor = today;
  if (!hitDays.has(cursor)) cursor = shiftDayKey(cursor, -1);
  while (hitDays.has(cursor)) {
    current += 1;
    cursor = shiftDayKey(cursor, -1);
  }

  return { current, longest };
}
